### ISSUE TITLE
Unit worker crash in musl libc with large response body

### ISSUE THREAD

**Issue by alexhoulton** [netbox-community/netbox-docker#539]
## Current Behaviour

We have a particularly heavy API call, that fails and causes the unitd worker to core dump on ~40% of requests.

https://netbox.corp/api/dcim/devices/?format=json&limit=0&role=role-1&role=role-2&role=role-3&role=role-4&role=role-5

This API request returns a Content-Length: 4039176 (around 4MB).

On failed requests, we see the following in the netbox container stdout:
```
user@netbox:/var/log$ sudo docker logs netbox_netbox_1 --follow | grep "core dumped"
2021/06/28 23:54:36 [alert] 1#1 process 31365 exited on signal 7 (core dumped)
```
Compiling a custom netbox image with `unitd-debug` present yields logs that show the response is generated correctly but then when attempting to send the response, we crash. Full log here: https://pastebin.com/h4Z2FarB but the following lines show the generation of response complete and prepared to send: 

```
2021/07/01 13:37:37.136 [debug] 733#733 [unit] outgoing plain buffer allocation: (0x7fb9248f5dd0, 387)
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: add_field #0: Content-Type: application/json
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: add_field #1: Vary: Accept, Cookie, Origin
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: add_field #2: Allow: GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: add_field #3: API-Version: 2.11
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: add_field #4: X-Content-Type-Options: nosniff
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: add_field #5: Referrer-Policy: same-origin
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: add_field #6: X-Frame-Options: SAMEORIGIN
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: add_field #7: Content-Length: 4039176
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: write: 4039176
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: send: 8 fields, 387 bytes
2021/07/01 13:37:37.136 [debug] 733#733 [unit] #271: send plain: 403
2021/07/01 13:37:37.136 [debug] 733#733 [unit] port{20,1} enqueue 1 read_socket notify 1
2021/07/01 13:37:37.136 [debug] 733#733 [unit] sendmsg(15, 403): 403
2021/07/01 13:37:37.136 [debug] 733#733 [unit] port{20,1} sendmsg 403
```

Going a step further, and enabling core dumps on the host, we can get the following from gdb:
```
bash-5.1$ gdb /usr/sbin/unitd /cores/core.unitd.250
GNU gdb (GDB) 10.1
Copyright (C) 2020 Free Software Foundation, Inc.
<snip>
[New LWP 250]
Core was generated by `unit: "netbox" application                                                    '.
Program terminated with signal SIGBUS, Bus error.
#0  0x00007fdab86b966b in memcpy () from /lib/ld-musl-x86_64.so.1
(gdb) quit
```

Which suggests the crash is in musl libc. SIGBUS seems to indicate the worker is trying to access memory that has not been physically mapped. 

Testing a theory, and building a custom netbox image from python:3.7-buster, we do not see the same crash. Likely as debian uses glibc rather than musl.

## Expected Behaviour

The API response is successful every time.

Nginx unit I note doesn't have official binaries for alpine linux, which suggests they may not have tested it. Could a solution be to provide netbox docker images based on debian or ubuntu, or another dist that uses glibc?

## Debug Information

I can provide core dumps if needed.

I'll also note that we saw a ~15% performance improvement from the buster based container over the alpine one, which is interesting more than anything, but strengthens the case for a debian based image build. (see image, real data from our production environment)

<img width="993" alt="Screenshot 2021-07-02 at 16 01 06" src="https://user-images.githubusercontent.com/298057/124303222-17eed600-db5a-11eb-9fd7-76ffc97fae64.png">

The output of `docker-compose version`: `docker-compose version 1.26.2, build unknown
docker-py version: 4.2.2
CPython version: 3.6.9
OpenSSL version: OpenSSL 1.1.1  11 Sep 2018`
The output of `docker version`: `Client: Docker Engine - Community
 Version:           20.10.5
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        55c4c88
 Built:             Tue Mar  2 20:18:05 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.5
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       363e9a8
  Built:            Tue Mar  2 20:16:00 2021
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          1.4.4
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
`
The output of `git rev-parse HEAD`: `676ca798cd684ce30fca2692180e680d2becf76e`
The command you used to start the project: `sudo docker-compose up -d`

The output of `cat docker-compose.override.yml`:
https://gist.github.com/madnutter56/8e60f106b5483751573aa6b7eee85336

**Comment by maximumG** [comment]
@madnutter56 I faced exactly the same issue with the same symptoms. 

Polling Netbox API devices endpoint leads to the CGI worker being killed with signal 7 in some specific cases with a large payload. On our side we started facing the issue when adding a lot of big config context associated to every devices.

We tried to limit the number of devices per requests (~500) but at some point the worker gets killed anyway when executing several queries one after another. The same nginx worker might handle the requests and get its memory overwhelmed after treating several request.

Furthermore, pynetbox library is causing the worker being killed every time, because pynetbox is requesting endpoint with args `/?limit=0`. We the pynetbox behavior by setting the MAX_PAGE_SIZE to 500 items max on netbox side. As a result even though you ask for limit=0 you end up every time with a maximum of 500 items.
